diff --git a/video/out/vulkan/context_moltenvk.m b/video/out/vulkan/context_moltenvk.m
index 445b907f7..90cadf957 100644
--- a/video/out/vulkan/context_moltenvk.m
+++ b/video/out/vulkan/context_moltenvk.m
@@ -27,6 +27,8 @@
 struct priv {
     struct mpvk_ctx vk;
     CAMetalLayer *layer;
+    int width;
+    int height;
 };

 static void moltenvk_uninit(struct ra_ctx *ctx)
@@ -77,13 +79,35 @@ static bool moltenvk_reconfig(struct ra_ctx *ctx)
 {
     struct priv *p = ctx->priv;
     CGSize s = p->layer.drawableSize;
-    ra_vk_ctx_resize(ctx, s.width, s.height);
+    int width = s.width;
+    int height = s.height;
+    if (!ra_vk_ctx_resize(ctx, width, height))
+        return false;
+    p->width = width;
+    p->height = height;
     return true;
 }

 static int moltenvk_control(struct ra_ctx *ctx, int *events, int request, void *arg)
 {
-    return VO_NOTIMPL;
+    if (request != VOCTRL_CHECK_EVENTS)
+        return VO_NOTIMPL;
+
+    struct priv *p = ctx->priv;
+    CGSize s = p->layer.drawableSize;
+    int width = s.width;
+    int height = s.height;
+    if (width <= 1 || height <= 1 ||
+        (width == p->width && height == p->height))
+        return VO_TRUE;
+
+    if (!ra_vk_ctx_resize(ctx, width, height))
+        return VO_ERROR;
+
+    p->width = width;
+    p->height = height;
+    *events |= VO_EVENT_RESIZE;
+    return VO_TRUE;
 }

 const struct ra_ctx_fns ra_ctx_vulkan_moltenvk = {
